services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientUpdateSourceHostsCondition.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.services.clientpolicy.condition;
+
+import java.util.Collections;
+import java.util.List;
+
+import org.jboss.logging.Logger;
+import org.keycloak.component.ComponentModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.services.clientpolicy.ClientPolicyContext;
+import org.keycloak.services.clientpolicy.ClientPolicyException;
+import org.keycloak.services.clientpolicy.ClientPolicyLogger;
+import org.keycloak.services.clientpolicy.ClientPolicyVote;
+
+public class ClientUpdateSourceHostsCondition implements ClientPolicyConditionProvider {
+
+    private static final Logger logger = Logger.getLogger(ClientUpdateSourceHostsCondition.class);
+
+    private final KeycloakSession session;
+    private final ComponentModel componentModel;
+
+    public ClientUpdateSourceHostsCondition(KeycloakSession session, ComponentModel componentModel) {
+        this.session = session;
+        this.componentModel = componentModel;
+    }
+
+    @Override
+    public String getName() {
+        return componentModel.getName();
+    }
+
+    @Override
+    public String getProviderId() {
+        return componentModel.getProviderId();
+    }
+
+    @Override
+    public ClientPolicyVote applyPolicy(ClientPolicyContext context) throws ClientPolicyException {
+        switch (context.getEvent()) {
+        case REGISTER:
+        case UPDATE:
+            if (isHostMatched()) return ClientPolicyVote.YES;
+            return ClientPolicyVote.NO;
+        default:
+            return ClientPolicyVote.ABSTAIN;
+        }
+    }
+
+    private boolean isHostMatched() {
+        String host = session.getContext().getRequestHeaders().getHeaderString("Host");
```

Comment:
I think that "Host" headers refers to the server host, rather than the host of the client.

I think you can take a look for the inspiration to the TrustedHostClientRegistrationPolicy . That one also provides some additional flexibility (EG. checking the domains instead of the pure hosts) and hence I suggest to just re-use some code from it to have similar flexibility here (also considering that we're going to replace "Client Registration Policies" by "Client Policies", we may need at least same flexibility, which is provided by the existing client registration policies)

Reply:
Also question is, if to "merge" this with the existing "Client IP" condition and just have same condition for both? So that admin can choose if he wants to use IP Address or the hostname? AFAIR this is supported by the TrustedHostClientRegistrationPolicy

Reply:
@mposolda I see, I agree with you. I would like to re-implement this condition to integrate the following 3 condition that has been planned originally.

- [KEYCLOAK-14197 Client Policy - Condition : Client - Client Host](https://issues.redhat.com/browse/KEYCLOAK-14197)
- [KEYCLOAK-14198 Client Policy - Condition : Client - Client IP](https://issues.redhat.com/browse/KEYCLOAK-14198) (I'll delete already merged this condition)
- [KEYCLOAK-14194 Client Policy - Condition : Client - Client Domain Name](https://issues.redhat.com/browse/KEYCLOAK-14194)

When re-implementing this condition, I will re-use codes from `TrustedHostClientRegistrationPolicy`.

WDYT?

Reply:
@tnorimat +1, Thanks!